#
#
#

import os
import sys
import logging

gitpath=os.path.expanduser("~/git/xskew")
sys.path.append(gitpath)

species='Homo_sapiens'
datadir = os.path.expanduser(f'~/data/cross_mammal_xci/temp/sra/{species}')
homedir = os.path.expanduser("~/")
wdir = f"{homedir}work/xskew1"
chr="X"

from xskew.tools import *
setup_logging(logging.DEBUG)

logging.debug('in snakefile...')

(SAMPLE,) = glob_wildcards(datadir + "/{sample}.sra")

logging.debug(f"HOME={homedir} WORKDIR={wdir} CHR={chr}")
logging.debug(f"SAMPLE={SAMPLE}")


rule all:
    input:
        # testing. 
        end1 = expand(wdir +  "/{sample}_1.fastq", sample=SAMPLE ),
        end2 = expand(wdir +  "/{sample}_2.fastq", sample=SAMPLE ),
        
        # from STAR
        abam= expand(wdir + "/{sample}.Aligned.out.bam", sample=SAMPLE ),
        
        
        # from STAR
        #log = expand(wdir + "/{sample}.Log.wasp.out", sample=SAMPLE ),        
        #flog= expand(wdir +"/{sample}.Log.final.wasp.out", sample=SAMPLE),
        #rpg=  expand(wdir +"/{sample}.ReadsPerGene.out.wasp.tab",    sample=SAMPLE)        

        # from gatk
        #sfvcf = expand(wdir + "/{sample}." + chr + ".snps_filtered.vcf",  sample=SAMPLE ),
        
        # from igvtools
        #wig =   expand(wdir + "/{sample}." + chr + ".split.filtered.wasp.wig", sample=SAMPLE )

rule dump_fastq:
    input:
        sra = datadir + "/{sample}.sra"
    
    output:
        end1= wdir + "/{sample}_1.fastq",    
        end2= wdir + "/{sample}_2.fastq"    

    resources:
        mem_mb=2048
    threads: 6

    run:
        fasterq_dump( infile=input.sra, 
                      outdir=wdir, 
                      nthreads=f"{threads}", 
                      tempdir ="./")
        
        
rule star_nowasp:
    input:
        end1= wdir + "/{sample}_1.fastq",   
        end2= wdir + "/{sample}_2.fastq"
    
    output:     
        abam= wdir + "/{sample}.Aligned.out.bam",
        rpg= wdir + "/{sample}.ReadsPerGene.out.tab",
        sjtab= wdir + "/{sample}.SJ.out.tab",
        log= wdir + "/{sample}.Log.out",        
        plog= wdir + "/{sample}.Log.progress.out",
        flog= wdir + "/{sample}.Log.final.out",
            
    params:        
        tmpdir = "{sample}_starout",
        wdir = wdir + "/",
        gdir= homedir + "data/genomes/GRCh38_Gencode25.2.4.9a/",
        outprefix = wdir + "/{sample}."
     
    resources:
        mem_mb=3072
    threads: 16
    
    run:
        star_nowasp( end1=input.end1, 
                     end2=input.end2, 
                     outprefix=params.outprefix,
                     outtemp= params.tmpdir,  
                     nthreads=f"{threads}", 
                     genomedir=params.gdir)
